### ISSUE TITLE
Root-implicit multisite has no energy objective: the cell engine differentiates a one-site observable, so it cannot drive a ground-state optimizer

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#894]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

Found while wiring the root-implicit engines into `optimize_gs_ad` (#715 follow-up, PR for the symmetric arm). The multisite (Phase 2) engine cannot be wired as things stand, and the reason is not plumbing.

## The gap

`cell_root_implicit_energy_and_grad` (`_ctm_root_implicit_multisite.py:957`) takes a second positional argument named `op` and differentiates

```python
rho = _rdm_1site_tensor(A_live, _to_ctm_env(env, template))
return jnp.real(jnp.trace(rho @ op))          # _cell_observable, :904
```

— a **one-site observable**, not an energy. Measured: feeding a two-site Heisenberg gate as `op` fails outright, because the RDM is `(2, 2)` and `rho @ gate.reshape(4, 4)` raises

```
dot_general requires contracting dimensions to have the same shape, got (2,) and (4,)
```

`_cell_energy` / `cell_energy_forward` do exist, and their own docstrings say why they are not the answer (`:928-941`):

> **Valid only on a uniform (1x1) cell.** `compute_energy_ctm_tensor` places the same `A` on both sites of a two-site RDM and glues them through one site's environment ring. On a cell of different tensors the two halves meet on chi bonds carrying independent gauges, so the result is gauge dependent and not a differentiable function of `A`.

That is measured, not argued (`_cell_observable`, `:886-892`): the value moves over `[0.143, 0.171]` under `|t| <= 2e-5` along a line while every fixed point converges to 8.6e-13, and finite differences diverge as `h -> 0`.

So the one-site observable is the *right* objective for the Phase 2 parity gate — it closes on the eight tensors of `env_ring_for_cell` and nothing else, so every bond gauge cancels — and it is the *wrong* objective for a ground state. `_cell_observable`'s docstring states the missing piece outright:

> a physical multisite *energy* needs a two-site ring spanning two cells and is separate work.

## What is needed

1. A two-site RDM whose ring spans **adjacent cells**, so both halves are closed by the same gauge — the multisite analogue of `_rdm2x1_tensor` / `_rdm1x2_tensor`.
2. Its adjoint, i.e. it has to be differentiable through the root-implicit variables the way `_cell_observable` is.
3. An FD parity gate on a genuinely non-uniform cell, since a uniform cell cannot distinguish the new energy from the existing (wrong-on-non-uniform) one.
4. Bond-counting that matches the 2-site checkerboard convention — cf. #494, where the 2-site bipartite energy was under-counting bonds.

## Everything else is genuinely plumbing (measured, so it is not the blocker)

For the record, since the wiring investigation covered it:

- the 2-tuple return contract already matches (`return_diagnostics=False`);
- optax handles `dict[(r,c) -> array]` params — `init`/`update`/`apply` all preserve keys, and `clip_by_global_norm` then clips the joint norm across sites, the same semantics as the existing tuple-params 2-site path;
- `_normalize_params` and the NaN-mask block need dict support (`jnp.linalg.norm(dict)` and `jnp.isfinite(dict)` both raise today); `_grad_l2_norm` already works;
- `iPEPSConfig` has no `nrows`/`ncols`, and `unit_cell="2site"` is a 2-coordinate `CHECKERBOARD_NEIGHBORS` map rather than a rectangular cell — a checkerboard needs `nrows=ncols=2` with four dict entries and **parameter tying** (two optax leaves fanned out to four engine entries, gradients summed back), because a 1x2 cell tiles as stripes;
- `rel_floor` is not a parameter of the cell engine at all;
- the final environment is fine: `python_loop_ctm_converge` + `compute_energy_ctm_tensor_multisite` both work on a 2x2 cell (measured, 33.9 s).

## Cost, for scoping

Measured wall time for one `cell_root_implicit_energy_and_grad` at D=2, χ=4 (CPU, after #731):

| cell | wall | root residual | adjoint residual |
|---|---|---|---|
| 1×1 | 56.9 s | 5.68e-14 | 1.61e-15 |
| 1×2 | 170.2 s | 1.15e-13 | 9.05e-14 |
| 2×2 | 510.4 s | 4.70e-13 | 2.85e-11 |

So any multi-step optimizer test at 2×2 is `slow`-bucket only, at ~8.5 min per gradient.

## Until then

`optimize_gs_ad_root_implicit` raises `NotImplementedError` for `unit_cell != "1x1"` and now names this as the reason, rather than the previous "built and tested but not yet wired here", which read as plumbing.

Refs #715, #494.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Scope: the multisite two-site energy objective for root-implicit AD

Investigated the adjoint structure of `cell_root_implicit_energy_and_grad` and the production multisite energy. **The scope is smaller than "physics vs plumbing" suggests, because the correct two-site energy already exists and is already differentiable** — the work is to route the root-implicit *regular* environment into it, not to re-derive a two-site ring.

### The one load-bearing observation

The whole adjoint apparatus in `cell_root_implicit_energy_and_grad` (`:1074-1136`) differentiates a single closure:

```python
def energy_of(a_data, corners_t, edges_t, S_all):
    c_reg, e_reg = absorb_inverse_roots_multisite(corners_t, edges_t, S_all, ...)
    A_live = DenseTensor(a_data[objective_cell], indices[objective_cell])
    return _cell_observable(A_live, c_reg, e_reg, template, op, objective_cell, ...)
```

`jax.vjp(energy_of, A_data, corners, edges, S)` produces both `grad_direct` (∂E/∂A) and `y_bar` (the cotangents on the root variables); everything after — `F_of_y`, `F_of_p`, `_solve_root_adjoint`, the `grad_direct − vjp_p(F_bar)` assembly — is **objective-agnostic and reused verbatim**. So the entire task reduces to replacing the body of `energy_of` with a differentiable two-site energy in the *same* variables `(a_data, corners, edges, S)`. No change to the root parametrization, the adjoint solve, the rank clamp, or the residual gate.

### What the new objective must be

`compute_energy_ctm_tensor_multisite(site_tensors, envs, neighbors, gate)` (`_ctm_tensor_energy.py:838`) is the reference. On a non-uniform cell it already:
- iterates each site's `right` + `bottom` bonds, de-duplicates via a `frozenset` bond id (so each bond is counted once — the #494 fix is inherited, not re-implemented),
- dispatches distinct-neighbor bonds to `_rdm2x1_tensor_2site` / `_rdm1x2_tensor_2site`, whose rings span two cells (`env_A` left/top, `env_B` right/bottom), and same-neighbor bonds to the single-A builders,
- returns `total / n_sites`, pure label contractions, **no SVD/eigh** → differentiable as-is.

Measured working on a non-uniform 2×2 (33.9 s, per this issue). So the objective is:

```python
def energy_of(a_data, corners_t, edges_t, S_all):
    c_reg, e_reg = absorb_inverse_roots_multisite(corners_t, edges_t, S_all, ...)
    a_live  = {rc: DenseTensor(a_data[rc], indices[rc]) for rc in a_data}   # all cells now
    envs    = {rc: _to_ctm_env(env_ring_for_cell(c_reg, e_reg, *rc, ...), template_rc)
               for rc in a_data}
    return compute_energy_ctm_tensor_multisite(a_live, envs, neighbors, gate)
```

Two consequences drop out for free: `grad_direct` becomes non-zero on **every** cell (a bond touches two sites), and `y_bar` touches more env tensors — `jax.vjp` handles both; the dict-valued cotangents simply have more non-zero entries. The gradient is then automatically correct **provided the assembled per-cell `CTMTensorEnv`s are gauge-consistent across the shared inter-cell bond** — which is the one real risk (below).

### The three concrete pieces of work

1. **Per-cell `CTMTensorEnv` assembly from the regular env** (`env_ring_for_cell` + `_to_ctm_env`, per coordinate, honoring the `above_left`/`above` shift convention already encoded at `:846`). `_cell_observable` already does this for one cell and is validated gauge-safe, so this is a loop over that existing bridge. A free internal check: the one-site RDM taken from the assembled `envs[objective_cell]` must reproduce today's `_cell_observable` value exactly.
2. **Swap `energy_of`'s body** to the block above and generalize the `y_bar` construction (the `u/v` cotangents stay zero; `c_bar/e_bar/S_bar` now come back non-zero on more coordinates — no structural change, the existing `jax.vjp` unpacking already returns full dicts).
3. **`template` per cell** — `energy_of` currently builds one `template` from `A_const[objective_cell]`; the multi-cell version needs one per coordinate (cheap; `initialize_ctm_tensor_env(A_const[rc], chi)`).

Everything the issue already flagged as plumbing (dict params through optax, `_normalize_params`/NaN-mask dict support, checkerboard parameter-tying `2×2 → two leaves`, the 2-tuple return) stays plumbing and lives in the **optimizer wrapper**, not in this objective.

### The one real risk: inter-cell gauge cancellation

The reason `_cell_observable` is smooth is that it closes on **one** ring, so every chi-bond gauge cancels internally. A two-site energy has an inter-cell bond (`t1_r↔t1_lR` for horizontal, `t4_u↔t4_dB` for vertical) whose gauge must cancel between **two separately assembled** rings. In production this holds because both per-site envs come from one converged fixed point. In the root-implicit path the regular env is reconstructed by `absorb_inverse_roots_multisite`, and the tilde tables **carry `S` between shifted cells** (`:982` docstring) — so the inter-cell gauge is threaded through `S` and *should* cancel by construction, the same way the one-site case does. **This is exactly what the FD-parity gate must prove**, and it is where months went in Phase 1 (#718, the `swap_env_convention` boundary). Budget the bulk of the effort here, not on the energy call.

### Validation plan (each guards a specific failure)

- **Internal consistency (cheap, `core`-able):** one-site RDM from the assembled `envs[cell]` == existing `_cell_observable` — proves the env assembly didn't perturb the gauge-safe ring.
- **FD parity on a genuinely NON-uniform cell (the gate):** a 2×2 with four distinct tensors (a uniform cell cannot distinguish the new energy from the old wrong-on-non-uniform one). **FD the sweep map from a fixed start, never the root** — `asym_root_parametrize` is gauge-discontinuous in p (memory: |ẏ_fd| diverges 1.98e4→1.19e5 though every point is a valid root). Assert the fixture is non-uniform enough that the *old* `cell_energy_forward` visibly fails on it, so the test can't go vacuous (#884 pattern).
- **Gradient vs production explicit-backprop** where the latter is finite (D=2), and **NaN-survival at D=3** — reproduce the Phase-1 headline (`test_the_gradient_survives_where_explicit_backprop_nans`) at multisite, since surviving degenerate spectra is the entire reason this path exists.
- **Gauge-consistency diagnostic** (`gauge_consistency`, already computed at `:1103-1108`) must stay ≤1e-8 with the two-site objective; if it doesn't, the inter-cell bond isn't cancelling.

### Cost / bucketing

Per this issue's table, one 2×2 gradient is ~510 s at D=2 χ=4 after #731, so every multi-step optimizer test is **`slow`-bucket only** (~8.5 min/gradient). The FD-parity and single-gradient checks are one gradient each; a convergence test is several — keep it to the minimum that demonstrates monotone descent on a non-uniform cell.

### Out of scope (state explicitly so it isn't assumed)

- **Fermionic** — same standing as the #986 CTM-AD NO-GO; multisite root-implicit stays dense/bosonic (`SymmetricTensor` input already raises `TypeError` at `:1004`).
- **Symmetric multisite** — a follow-on; this scope is the dense engine only.
- **Converged benchmark vs QMC/variPEPS** — separate deliverable; this issue closes when the objective is correct and FD-gated, not when a paper number is reproduced.

### Effort estimate

The energy swap + env assembly is ~1 day. The FD-parity gate on a non-uniform cell — deriving a fixture that exercises the inter-cell gauge and confirming cancellation — is the real cost and the #718-class risk; realistically 3–5 days including the convention debugging that boundary historically demands. Net: **a focused week**, with the risk concentrated entirely in the inter-cell gauge boundary, not in the energy itself.
